pkg/datapath/loader/netkit.go

```diff
@@ -0,0 +1,187 @@
+	}
+
+	l, err := link.AttachNetkit(link.NetkitOptions{
+		Program:   prog,
+		Attach:    attach,
+		Interface: device.Attrs().Index,
+		Anchor:    link.Tail(),
+	})
+	if err != nil {
+		return fmt.Errorf("attaching netkit: %w", err)
+	}
+
+	defer func() {
+		// The program was successfully attached using netkit. Closing
+		// a link does not detach the program if the link is pinned.
+		if err := l.Close(); err != nil {
+			log.Warnf("Failed to close netkit link for program %s", progName)
+		}
+	}()
+
+	pin := filepath.Join(bpffsDir, progName)
+	if err := l.Pin(pin); err != nil {
+		return fmt.Errorf("pinning link at %s for program %s : %w", pin, progName, err)
+	}
+
+	log.Infof("Program %s attached to device %s using netkit", progName, device.Attrs().Name)
+
+	return nil
+}
+
+// updateNetkit attempts to update an existing netkit link called progName
+// in bpffsDir. If the link is defunct, the pin is removed.
+//
+// Returns nil if the update was successful. Returns an error wrapping
+// [os.ErrNotExist] if the link is defunct or missing.
+func updateNetkit(prog *ebpf.Program, progName, bpffsDir string) error {
+	// Attempt to open and update an existing link.
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UpdateLink(pin, prog)
+	switch {
+	// Link exists, but is defunct, and needs to be recreated. The program
+	// no longer gets triggered at this point and the link needs to be removed
+	// to proceed.
+	case errors.Is(err, unix.ENOLINK):
+		if err := os.Remove(pin); err != nil {
+			return fmt.Errorf("unpinning defunct link %s: %w", pin, err)
+		}
+
+		log.Infof("Unpinned defunct link %s for program %s", pin, progName)
+
+		// Wrap in os.ErrNotExist so the caller needs to look for one error.
+		return fmt.Errorf("unpinned defunct link: %w", os.ErrNotExist)
+
+	// No existing link found, continue trying to create one.
+	case errors.Is(err, os.ErrNotExist):
+		log.Debugf("No existing link found at %s for program %s", pin, progName)
+		return err
+
+	case err != nil:
+		return fmt.Errorf("updating link %s for program %s: %w", pin, progName, err)
+	}
+
+	log.Infof("Updated link %s for program %s", pin, progName)
+	return nil
+}
+
+// detachNetkit attempts to open the link at progName in bpffsDir
+// and unpins it. Only returns unrecoverable errors. Returns nil if
+// the link doesn't exist or if removal was successful.
+func detachNetkit(bpffsDir, progName string) error {
+	pin := filepath.Join(bpffsDir, progName)
+	err := bpf.UnpinLink(pin)
+	if err == nil {
+		log.Infof("Removed netkit link at %s", pin)
+		return nil
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return nil
+	}
+
+	// The pinned link exists, something went wrong unpinning it.
+	return fmt.Errorf("unpinning netkit link: %w", err)
+}
+
+// hasCiliumNetkitLinks returns true if device has a Cilium-managed
+// netkit program with the given attach type.
+func hasCiliumNetkitLinks(device netlink.Link, attach ebpf.AttachType) (bool, error) {
```

Comment:
Nit: this function is almost exactly the same as `hasCiliumTCXLinks`. Maybe we can merge those two or factor out the similarities as a follow up.

Reply:
Yes, makes sense, I'll look into this as followup.